### PR TITLE
Sunrise 0.06

### PR DESCRIPTION
**Pull request by g-rden**
Had to change a lot, but I think this app is in a good state now.

### CODE REVIEW

File: apps/sunrise/app.js [hunk 1]
```diff
@@ -18,329 +17,206 @@ const latlon = loadLocation() || {};
...
+daystart.setHours(0);
+daystart.setMinutes(0);
+daystart.setSeconds(0);
 
-function formatAsTime (hour, minute) {
-  return '' + ((hour < 10) ? '0' : '') + (0 | hour) +
-         ':' + ((minute < 10) ? '0' : '') + (0 | minute);
+function getAltitude () {
+  Bangle.getPressure().then(d=>{
+    alt = d.altitude;
```
[the remark this answers is not in the document]
Comment (g-rden) [reply apps/sunrise/app.js:57]: Hmm, that reminds me. I ignored that the app also isn't working in the web emulator.
```
Uncaught Error: Cannot read property 'then' of undefined
 at line 1 col 21
Bangle.getPressure().then(d=>{
                    ^
```
So the whole thing could be handled better, but idk how.

**Comment by gfwilliams** [comment]
Have you done any benchmarking on this? My gut feeling is that using a lookup table for sin is probably not going to make much difference to speed (although it'll slow down the first load of the clock).

Plus you seem to have twice as many elements with the first of each pair being `sineLUT[i * 2] = i;` - when if you just used half the amount of elements, you could just use the index itself and avoid all the `*2+1` stuff you're having to do to access the other item?

**Comment by g-rden** [comment]
I didn't benchmark it and I didn't look into how to do that yet. My goal was to update variables only at first load and every new day. I am aware that it might not make a difference.

I don't get how there could be twice the elements in sineLUT than needed. sineLUT stores x and y values, therefor it needs to be twice the number of xy pairs. Unless you suggest halving the amount of xy pairs.

**Comment by gfwilliams** [comment]
My point was that your lut has one entry per pixel on the screen. You do:

```
  for (let i = 0; i < w; i++) {
    sineLUT[i * 2] = i;
    sineLUT[1 + i * 2] = ypos(i);
  }
...
   sinStep=20;
  for (i = 0; i <= sineLUT.length - sinStep; i += sinStep) {
    g.drawLine(sineLUT[i], sineLUT[1 + i],
               sineLUT[sinStep + i], sineLUT[sinStep + 1 + i]);
  }
```

When all you needed if you didn't store x was:

```
 sinStep=10;
  for (let i = 0; i < w; i++) 
    sineLUT[i] = ypos(i);
...
  for (i = 0; i < sineLUT.length*2; i += sinStep) {
    g.drawLine(i, sineLUT[i], sinStep + i, sineLUT[sinStep + i]);
  }
```

I'm not entirely sure I understand that code though as it looks a lot like sinStep=20 so you were calculating 10x more points in your lookup table than you ever use?

Another thing to mention is if you are storing x+y in an array, you can pass it straight into g.drawPoly with `g.drawPoly(sinLUT)` in which case it really will be very fast! That might be the best solution

**Comment by g-rden** [comment]
Now I get it. I didn't do that to be consistant. There are multiple places that look like that, which can't be written like you suggest. So I wrote it everywhere the same. Might not be the best way of doing it.

sineLUT is not only used for drawing the sine. It is also used for getting the points for sunrise, sunset and the sun position, which are all pixel accurate. The sineStep is for drawing fewer sine lines. It's value is rather arbitrary.

I experimented with drawPoly. But it can't draw every element: `Note: there is a limit of 64 points (128 XY elements) for polygons`. I didn't use it because I don't want to draw every element from sineLUT, thus there would need to be another array afaik. I can use drawPoly if it's faster.

File: apps/sunrise/app.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (bobrippling) [reply apps/sunrise/app.js:57]: Ah yes - the emulator doesn't have `getPressure` - we could do this:

```javascript
Bangle.getPressure && Bangle.getPressure().then(d => {
    d && (alt = d.altitude);
});
```

[the remark this answers is not in the document]
Comment (g-rden) [reply apps/sunrise/app.js:57]: I get errors writing it like this.

Is this ok?
```
  if (Bangle.getPressure()) {
    Bangle.getPressure().then(d => {
    if (d) alt = d.altitude;
    });
  }
```

[the remark this answers is not in the document]
Comment (bobrippling) [reply apps/sunrise/app.js:57]: Ah nice spot, yes - that's what we want, and it'll work but we can avoid creating the promise twice:
```javascript
const p = Bangle.getPressure();
if (p) {
  p.then(d => {
    if (d) alt = d.altitude;
  });
}
```

[the remark this answers is not in the document]
Comment (g-rden) [reply apps/sunrise/app.js]: Great, Thanks!

**Comment by bobrippling** [comment]
Sorry - I've caused these issues with [my PR](https://github.com/espruino/BangleApps/pull/3066). I've pushed the fix, hope you don't mind!

**Comment by g-rden** [comment]
I don't mind, but I call it quits and close this PR

**Comment by bobrippling** [comment]
@g-rden, I quite like the sunrise clockface, would you mind if I took these changes and brought them in?

**Pull request closed without merging**